### ISSUE TITLE
add dedicated warning code for relative import with no parent package

### ISSUE THREAD

**Issue by danielpopescu** [astral-sh/ty#2589]
### Summary

This likely already reported, but just in case, importing modules from the same directory as the current script/file is not working....

**Comment by MichaReiser** [comment]
This should be supported. Can you share an example of where it isn't? Otherwise, this issue isn't really actionable.

**Comment by danielpopescu** [comment]
Hmm... Im using latest ty 0.0.13, emacs, macos.

In a folder, say tmp, create a small file logger.py with this content (it doesnt matter what, should just have the class Logger in it):

```py
from __future__ import annotations

import logging


class Logger:

    @staticmethod
    def msg(msg: str) -> None:
        Logger._init()
        logging.getLogger(__name__).info(msg)

    # private
    _inited: bool = False

    @staticmethod
    def _init() -> None:
        if not Logger._inited:
            logging.basicConfig(format="%(message)s", level=logging.INFO)
            Logger._inited = True

```

then open a new file foo.py in same folder tmp and type:

```py
from .logger import Logger

```

I get the error in the attached 

![Image](https://github.com/user-attachments/assets/bcf2dede-74ad-4e92-b46e-2edb91877ae3)

screenshot.

**Comment by danielpopescu** [comment]
Note that ruff does solve the import file... (it just complains that is unused....)

**Comment by MichaReiser** [comment]
Thanks, that's already helpful context. This should definitely work. Would you mind setting [`logLevel`](https://docs.astral.sh/ty/reference/editor-settings/#initialization-options) to debug and sharing the logs with us?

**Comment by AlexWaygood** [comment]
This is probably because we're setting `tmp` as the first-party search path when the first-party search path would need to be `tmp/..` (the parent directory of `tmp`) in order for us to resolve `tmp` as a namespace package and resolve the relative import

**Comment by MichaReiser** [comment]
Ah yes. I always forget how relative imports work in Python. It's because our desperate module resolution only kicks in for folders with a `pyproject.toml`.

**Comment by George-Ogden** [comment]
I also have this issue. Here's an MRE:
`src/box.py`: `class Box: ...`
`src/box_test.py`: `from .box import Box`
`src/__init__.py` (empty)

**Comment by AlexWaygood** [comment]
@George-Ogden, that's not the conventional way to use an `src/` directory in Python -- see https://github.com/astral-sh/ty/issues/1398#issuecomment-3419883592

Edit: but maybe we can fix it for you anyway, it probably won't be too hard...

**Comment by evanjs** [comment]
Not entirely sure if mine is the _exact_ same issue, but it definitely feels like it

Basically, ComfyUI and extensions require the format noted by OP
That is to say, when extensions are loaded in ComfyUI, they _must_ use relative imports, or they won't work with how Comfy tries to load them

There are plenty of examples of such relative imports across the GH Org
https://github.com/search?q=org%3AComfy-Org+%2Ffrom+%5C..*%3F+import%2F+path%3A**.py&type=code

## Workaround (in the context of extensions)

To work around this, I need to ignore each import line that `ty` (and by extension, `ruff`) get angry about

### Example

```py
try:
    from .utils import *  # ty:ignore[unresolved-import] # noqa: F403
except ImportError:
    from utils import * # noqa: F403
```

(The `except` case above is just so the module can also be accessed outside of the Comfy ecosystem, e.g. CLI, testing, etc.)

With the above suppresions, both `ty` and `ruff` stop whining

**Comment by lampysprites** [comment]
Same issue here as the first post. I'm using system python environment, and all default config except ignoring 'invalid-type-form' and log level. Python scripts are all at top level and there's no subdirs. I tried creating pyproject.toml that sets root to the root dir to no effect. I also looked in FAQ and environment section of the docs.
Here's the logs from vscode: [ty.log](https://github.com/user-attachments/files/25080002/ty.log)
Here's one example of sources that fail [peel_25_12.zip](https://github.com/user-attachments/files/25080070/peel_25_12.zip)

* Edit: I'm in a similar situation as evanj just posted by the way. The scripts are executed by built-in interpreter in blender, and its import paths are a bit different.

**Comment by gordio** [comment]
Version: 0.15
Relative imports ([PEP 328](https://peps.python.org/pep-0328/#rationale-for-relative-imports)) still don't work.

![Image](https://github.com/user-attachments/assets/4c12b712-23f5-48f9-ab0e-88b6f6e95c90)

**Comment by carljm** [comment]
To be clear, ty definitely supports relative imports. If they aren't working in your specific case, that indicates that the root paths are not configured in such a way as to make those relative imports legal.

It looks like at least several of the reported cases here are attempts to relative-import something that ty considers a top-level module (because it's a file located directly in the root folder, not in a subfolder.) ty doesn't allow these relative imports because they aren't allowed at runtime, either:

```py
➜ ls
a.py    main.py

➜ cat a.py
A = 1

➜ cat main.py
from .a import A

➜ python3 main.py
Traceback (most recent call last):
  File "/Users/carlmeyer/tmp/main.py", line 1, in <module>
    from .a import A
ImportError: attempted relative import with no known parent package
```

But I think it is too confusing to simply not allow such imports; we probably need to detect the attempt and issue a specific warning about it, which can be disabled for the cases mentioned above where at runtime the import paths are different.

**Comment by carljm** [comment]
Taking the `server` tag off, since I think this issue can apply both to server and non-server contexts, and re-titling the issue.

**Comment by carljm** [comment]
We do already have this help message suggesting to remove the leading dot:

```py
➜ uvx ty check .
error[unresolved-import]: Cannot resolve imported module `.a`
 --> main.py:1:7
  |
1 | from .a import A
  |       ^
  |
help: The module can be resolved if the number of leading dots is reduced
help: Did you mean `a`?
info: Searched in the following paths during module resolution:
info:   1. /Users/carlmeyer/tmp (first-party code)
info:   2. vendored://stdlib (stdlib typeshed stubs vendored by ty)
info:   3. /Users/carlmeyer/.cache/uv/archive-v0/TS1i64PT4zMDifvqKGAbg/lib/python3.14/site-packages (site-packages)
info: make sure your Python environment is properly configured: https://docs.astral.sh/ty/modules/#python-environment
info: rule `unresolved-import` is enabled by default

Found 1 diagnostic
```

But that doesn't help if you are using a framework that imports your modules in a way where it requires the leading dot.

**Comment by carljm** [comment]
For reference, mypy also errors on such relative imports. Pyright and pyrefly allow them.
